### PR TITLE
Hide comment feature

### PR DESCRIPTION
**Pull request by Zhao-Andy**
## What type of PR is this? (check all applicable)
- [x] Feature

## Description
This PR gives authors of a post to hide comments in their post. Hidden comments will not be viewable in the main comment thread or any comment thread, but can be viewed in their respective permalink / their "show page." 

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)

example of hidden comments in a thread
example of hidden comments in a thread

example of dropdown of other person's comment
dropdown example of other comment

example of dropdown of your own comment
dropdown example of your own comment

example of hidden comment in permalink view
example of hidden comment in permalink view

## Added to documentation?
- [x] FAQ (to added when merged)

## [optional] What gif best describes this PR or how it makes you feel?

A cat hides itself inside a couch, as cats do.

### CODE REVIEW

File: app/controllers/comments_controller.rb [hunk 1]
```diff
@@ -153,6 +153,28 @@ def settings
...
+  end
+
+  def unhide
+    @comment = Comment.find(params[:comment_id])
+    authorize @comment # check for commentable author
+    @comment.hidden_by_commentable_user = false
+    if @comment.save
+      render json: { hidden: "false" }, status: :ok
+    else
+      render json: { errors: @comment.errors.full_messages.join(", "), status: 422 }, status: :unprocessable_entity
```
Comment (Zhao-Andy) [inline app/controllers/comments_controller.rb]: I'm not sure if rendering JSON here was the best implementation. It makes this controller not RESTful. Perhaps these actions belong in the API controller or something?

Also, the end result is reloading the page.

File: app/javascript/contentDisplayPolicy/initHiddenComments.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+      method: 'PATCH',
+      headers: {
+        'X-CSRF-Token': window.csrfToken,
+      },
+    })
+      .then(response => response.json())
+      .then(response => {
+        if (response.hidden === 'false') {
+          /* eslint-disable-next-line no-restricted-globals */
+          location.reload();
```
Comment (Zhao-Andy) [inline app/javascript/contentDisplayPolicy/initHiddenComments.js]: Using `location.reload()` feels like a hack to me.

File: app/views/comments/_comment_proper.html.erb [hunk 3]
```diff
@@ -25,40 +25,55 @@
...
+              </a>
+            <% end %>
             <span class="comment-actions hidden" data-user-id="<%= comment.user_id %>" style="display: none;">
               <a href="<%= comment.path %>/settings" rel="nofollow" style="color:#0a0a0a;" data-no-instant>Settings</a>
             </span>
+            <% action = comment.hidden_by_commentable_user ? "Unhide" : "Hide" %>
+            <span class="comment-actions hidden" data-action="hide-button" data-commentable-user-id="<%= commentable.user_id %>" data-user-id="<%= comment.user_id %>" style="display: none; width: 100%;">
+              <button class="hide-comment" type="button" data-hide-type="<%= action.downcase %>" data-comment-id="<%= comment.id %>">
+                <%= action %>
+              </button>
```
Comment (Zhao-Andy) [inline app/views/comments/_comment_proper.html.erb]: Another weird implementation thing I have about this: it's a button that sends a `fetch` request, that then runs `location.reload()` on success. I feel like there's a more "Rails" way of resolving this, like making this a `<form>` that sends a regular `PATCH` request. I tried for a few minutes to get the `<form>` to work but the CSS was messy and the form didn't work as well anyway.

File: db/migrate/20191108153914_add_hidden_by_commentable_user_to_comments.rb [hunk 4]
```diff
@@ -0,0 +1,5 @@
+class AddHiddenByCommentableUserToComments < ActiveRecord::Migration[5.2]
+  def change
+    add_column :comments, :hidden_by_commentable_user, :boolean, default: false
```
Comment (Zhao-Andy) [inline db/migrate/20191108153914_add_hidden_by_commentable_user_to_comments.rb]: Probably doesn't need an index right?

Comment (benhalpern) [reply db/migrate/20191108153914_add_hidden_by_commentable_user_to_comments.rb]: This is correct. If we had a `where(hidden_by_commentable_user: true)` call somewhere we might want one.

File: app/views/comments/_comment.html.erb [hunk 5]
```diff
@@ -1,5 +1,5 @@
 <% if comment && comment.user %>
-  <% if comment.depth < 3 %>
+  <% if comment.depth < 3 && !(should_be_hidden?(comment, @root_comment) && comment.depth.zero?) %>
```
Comment (benhalpern) [inline app/views/comments/_comment.html.erb]: I'm seeing top level comments here when they should be hidden altogether

Screen Shot 2019-12-02 at 5 15 52 PM

File: app/views/comments/index.html.erb [hunk 6]
```diff
@@ -98,14 +98,22 @@
...
       <% Comment.tree_for(@commentable).each do |comment, sub_comments| %>
+      <% any_comments_hidden = true if comment.hidden_by_commentable_user %>
         <% cache ["comment_root_#{user_signed_in?}", comment] do %>
           <%= tree_for(comment, sub_comments, @commentable) %>
         <% end %>
       <% end %>
     <% end %>
   </div>
   <center style="margin:30px 0px 22px;">
+  <% if any_comments_hidden %>
```
Comment (benhalpern) [inline app/views/comments/index.html.erb]: This logic doesn't seem to be working for me

Screen Shot 2019-12-02 at 5 17 46 PM

File: app/views/comments/_comment_proper.html.erb [hunk 7]
```diff
@@ -25,40 +25,55 @@
...
-            </a>
+            <% unless should_be_hidden?(comment, @root_comment) %>
+              <a href="<%= comment.path %>">
+                Permalink
+              </a>
+            <% end %>
             <span class="comment-actions hidden" data-user-id="<%= comment.user_id %>" style="display: none;">
               <a href="<%= comment.path %>/settings" rel="nofollow" style="color:#0a0a0a;" data-no-instant>Settings</a>
             </span>
+            <% action = comment.hidden_by_commentable_user ? "Unhide" : "Hide" %>
```
Comment (benhalpern) [inline app/views/comments/_comment_proper.html.erb]: Even as not-the-author I'm able to see a dropdown with the hide option on a hidden comment.

Screen Shot 2019-12-02 at 5 18 35 PM

File: app/assets/stylesheets/comments.scss [hunk 8]
```diff
@@ -734,6 +742,11 @@ a.header-link {
         }
       }
     }
+    &.hidden-comment {
```
Comment (benhalpern) [inline app/assets/stylesheets/comments.scss]: I don't think this should be faded. I feel like having it similar to the community standards violation design here could shame someone in the event of a hidden comment, which isn't what we're looking to do here.

File: app/javascript/contentDisplayPolicy/initHiddenComments.js (hunk 2 above)
Comment (benhalpern) [reply app/javascript/contentDisplayPolicy/initHiddenComments.js]: We could either disappear the comment (if it's top level) and replace the comment with the [hidden] version if not. But I don't think `location.reload()` is so bad either. This won't be used super often and reloading is a clean way to ensure it works properly.

**Comment by Zhao-Andy** [comment]
Made some commits; will fix the top level issue tomorrow. I think that was a dangling functionality that I forgot to note after our sync about it.

**Comment by benhalpern** [comment]
Padding is jarred a bit here on the hidden permalink

Screen Shot 2019-12-04 at 8 50 15 AM

**Comment by benhalpern** [comment]
Looking pretty close!

I'm seeing that the below message is sometimes lost beneath what's above it due to a negative margin. Not entirely sure why, but needs to be explored.

Screen Shot 2019-12-05 at 5 43 36 PM

Also looks like the permalink is no longer working properly...

Screen Shot 2019-12-05 at 5 45 09 PM

**Comment by Zhao-Andy** [comment]
@benhalpern okay this should be pretty good now. I fixed a merge conflict that will probably make the `comment_policy_spec.rb` fail, and can look at that a bit later.

**Comment by benhalpern** [comment]
I realized the caching logic could mess with the way we declare whether any comments have been hidden so I added a dedicated column which will clean up the view logic. Just trying to push this over the finish line.

**Pull request merged**